drivers/i2c/i2c_sam_twihs_rtio.c

```diff
@@ -0,0 +1,450 @@
+
+	i2c_sam_twihs_start(dev);
+}
+
+static void i2c_sam_twihs_complete(const struct device *dev, int status)
+{
+	const struct i2c_sam_twihs_dev_cfg *const dev_cfg = dev->config;
+	struct i2c_sam_twihs_dev_data *const dev_data = dev->data;
+	Twihs *const twihs = dev_cfg->regs;
+	struct rtio_iodev_sqe *iodev_sqe = dev_data->iodev_sqe;
+
+	/* Disable all enabled interrupts */
+	twihs->TWIHS_IDR = twihs->TWIHS_IMR;
+
+	if (status < 0) {
+		rtio_iodev_sqe_err(iodev_sqe, status);
+		i2c_sam_twihs_next(dev, true);
+		return;
+	}
+
+	if (dev_data->sqe->flags & RTIO_SQE_TRANSACTION) {
+		dev_data->sqe = rtio_spsc_next(iodev_sqe->r->sq, dev_data->sqe);
+		i2c_sam_twihs_start(dev);
+	} else {
+		rtio_iodev_sqe_ok(iodev_sqe, status);
+		i2c_sam_twihs_next(dev, true);
+	}
+}
+
+static void i2c_sam_twihs_submit(const struct device *dev, struct rtio_iodev_sqe *iodev_sqe)
+{
+	struct i2c_sam_twihs_dev_data *const dev_data = dev->data;
+
+	rtio_mpsc_push(&dev_data->io_q, &iodev_sqe->q);
+	i2c_sam_twihs_next(dev, false);
+}
+
+static void i2c_sam_twihs_isr(const struct device *dev)
+{
+	const struct i2c_sam_twihs_dev_cfg *const dev_cfg = dev->config;
+	struct i2c_sam_twihs_dev_data *const dev_data = dev->data;
+	Twihs *const twihs = dev_cfg->regs;
+	const struct rtio_sqe *const sqe = dev_data->sqe;
+	uint32_t isr_status;
+
+	/* Retrieve interrupt status */
+	isr_status = twihs->TWIHS_SR & twihs->TWIHS_IMR;
+
+	/* Not Acknowledged */
+	if (isr_status & TWIHS_SR_NACK) {
+		i2c_sam_twihs_complete(dev, -EIO);
+		return;
+	}
+
+	/* Byte received */
+	if (isr_status & TWIHS_SR_RXRDY) {
+		sqe->buf[dev_data->buf_idx] = twihs->TWIHS_RHR;
+		dev_data->buf_idx += 1;
+
+		if (dev_data->buf_idx == sqe->buf_len - 1U) {
+			/* Send STOP condition */
+			twihs->TWIHS_CR = TWIHS_CR_STOP;
+		}
+	}
+
+	/* Byte sent */
+	if (isr_status & TWIHS_SR_TXRDY) {
+		if (dev_data->buf_idx == sqe->buf_len) {
+			if (sqe->iodev_flags & RTIO_IODEV_I2C_STOP) {
+				/* Send STOP condition */
+				twihs->TWIHS_CR = TWIHS_CR_STOP;
+				/* Disable Transmit Ready interrupt */
+				twihs->TWIHS_IDR = TWIHS_IDR_TXRDY;
+			} else {
+				/* Transmission completed */
+				i2c_sam_twihs_complete(dev, 0);
+				return;
+			}
+		} else {
+			twihs->TWIHS_THR = sqe->buf[dev_data->buf_idx++];
+		}
+	}
+
+	/* Transmission completed */
+	if (isr_status & TWIHS_SR_TXCOMP) {
+		i2c_sam_twihs_complete(dev, 0);
+	}
+}
+
+static int i2c_sam_twihs_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+				  uint16_t addr)
+{
+	struct i2c_sam_twihs_dev_data *const dev_data = dev->data;
+	struct rtio_iodev *iodev = &dev_data->iodev;
+	struct rtio *const r = dev_data->r;
+	struct rtio_sqe *sqe = NULL;
+	struct rtio_cqe *cqe = NULL;
+	int res = 0;
+
+	k_sem_take(&dev_data->block_lock, K_FOREVER);
+
+	dev_data->dt_spec.addr = addr;
+
+	sqe = i2c_rtio_copy(r, iodev, msgs, num_msgs);
+	if (sqe == NULL) {
+		LOG_ERR("Not enough submission queue entries");
+		res = -ENOMEM;
```

Comment:
is there a possibility to improve that in future? Handling in chunks. Copy only first complete sequence (e.g. TX-RX) wait for completion copy next one so that queue must be of size of the biggest possible sequence.

Reply:
Yes, though at least a pair needs to be completed per transaction presumably for the common case